[Cocoa] Cancel pending getDisplayMedia prompt when page is reloaded

eric-carlson opened the pull request:
#### 80f25d4770c9630760067facda8dd3a670784e5b
<pre>
[Cocoa] Cancel pending getDisplayMedia prompt when page is reloaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=260795">https://bugs.webkit.org/show_bug.cgi?id=260795</a>
<a href="https://rdar.apple.com/114563662">rdar://114563662</a>

Reviewed by Youenn Fablet.

If the page is reloaded while the system prompt window/screen picker is active, cancel
the picker instead waiting for the ScreenCaptureKitSharingSessionManager watchdog timer
to go off and cancel it.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
(WebCore::ScreenCaptureKitSharingSessionManager::cancelGetDisplayMediaPrompt):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::cancelGetDisplayMediaPrompt):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::cancelGetDisplayMediaPrompt):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::cancelGetDisplayMediaPrompt):
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.h:
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
(WebKit::UserMediaPermissionRequestProxyMac::invalidate):
(WebKit::UserMediaPermissionRequestProxyMac::promptForGetDisplayMedia):

Canonical link: <a href="https://commits.webkit.org/274439@main">https://commits.webkit.org/274439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07881eb0fc402915373bdc334fc818afe65cba9c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17145 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18166 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19857 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16854 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17752 "14 flakes 165 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16586 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17022 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14176 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14265 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14859 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15252 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15024 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21660 "2 flakes 103 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13262 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19335 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14819 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8748 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15433 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -332,6 +332,10 @@ void UserMediaPermissionRequestManagerProxy::resetAccess(std::optional<FrameIden
+        if (m_currentUserMediaRequest && m_currentUserMediaRequest->frameID() == *frameID) {
+            m_currentUserMediaRequest->invalidate();
+            m_currentUserMediaRequest = nullptr;
+        }
```
Should we also invalidate m_currentUserMediaRequest if there is no frameID?

Something like below out of the if(frameID):
```
if (m_currentUserMediaRequest && (!frameID || m_currentUserMediaRequest->frameID() == *frameID)) {
...
}
```

eric-carlson replied:
Yes, good idea.

youennf commented on Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.h:
```diff
@@ -36,8 +36,12 @@ class UserMediaPermissionRequestProxyMac final : public UserMediaPermissionReque
     bool canRequestDisplayCapturePermission() final;
+    void invalidate() final;
+
+    bool m_pendingGetDispayMediaPrompt { false };
```
s/m_pendingGetDispayMediaPrompt/m_hasPendingGetDispayMediaPrompt.

eric-carlson replied:
Changed

youennf commented on Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
```diff
@@ -47,14 +47,26 @@
+    m_pendingGetDispayMediaPrompt = true;
     DisplayCaptureSessionManager::singleton().promptForGetDisplayMedia(promptType, Ref { manager()->page() }, topLevelDocumentSecurityOrigin().data(), [protectedThis = Ref { *this }](std::optional<CaptureDevice> device) mutable {
 
+        protectedThis->m_pendingGetDispayMediaPrompt = false;
```
if m_pendingGetDispayMediaPrompt is false, we could probably safely return early?

eric-carlson replied:
It will be false if `invalidate()` has been called, but we still want to reject the promise in that case.

eric-carlson committed:
[Cocoa] Cancel pending getDisplayMedia prompt when page is reloaded
https://bugs.webkit.org/show_bug.cgi?id=260795
rdar://114563662

Reviewed by Youenn Fablet.

If the page is reloaded while the system prompt window/screen picker is active, cancel
the picker instead waiting for the ScreenCaptureKitSharingSessionManager watchdog timer
to go off and cancel it.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
(WebCore::ScreenCaptureKitSharingSessionManager::cancelGetDisplayMediaPrompt):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::cancelGetDisplayMediaPrompt):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::cancelGetDisplayMediaPrompt):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::cancelGetDisplayMediaPrompt):
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.h:
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
(WebKit::UserMediaPermissionRequestProxyMac::invalidate):
(WebKit::UserMediaPermissionRequestProxyMac::promptForGetDisplayMedia):

Canonical link: https://commits.webkit.org/274439@main
